### DISCUSSION TITLE
[show-and-tell] {targets} and {condathis}: R pipelines Using R and Other CLI Tools

### DISCUSSION THREAD

**Discussion by c1au6i0** [ropensci/targets#1406] [Show and tell]
In our bioinformatics lab, we love using `targets` for omics analysis. One challenge we faced was that many upstream steps in the analysis require other, non-`R` command-line interface (CLI) tools. To allow us to create `targets` reproducible pipelines that integrate both `R` and other CLI tools within a full `R` environment, my friend @luciorq created a CRAN package called `condathis`. This package was inspired by our experience with `targets`, and we wanted to share it here to show our appreciation for the work done on it.  👏👏

If you are curious, check out the [condathis repository](https://github.com/luciorq/condathis) for more details. 
Here’s also a **practical [example](https://github.com/c1au6i0/align-condathis-targets) of a  reproducible pipeline** that employs `targets` and `condathis` to run common bioinformatic CLI tools for the in initial steps of an omics analysis.

**Comment by wlandau** [comment]
That's awesome, thanks for sharing `condathis`! FYI @joelnitta.

**Reply by joelnitta** [reply]
Cool! I had a similar idea [to do this with docker containers](https://www.joelnitta.com/talks/2023-01-01_ropensci).

**Reply by c1au6i0** [reply]
Thank you, will check that out @joelnitta :)

**Reply by luciorq** [reply]
Thank you, @c1au6i0 and @joelnitta,

I really enjoyed your [rOpenSci presentation](https://www.joelnitta.com/talks/2023-01-01_ropensci
) and the examples you shared, such as [targets_vcf_example](https://github.com/joelnitta/targets_vcf_example). They’re excellent illustrations of tackling a really common challenge in tool management.

With [condathis](https://github.com/luciorq/condathis), my primary goal was to adopt an “Environment as Code” approach for managing external tools while offering a familiar API akin to `processx::run()`. It also aims to improve portability, particularly for HPC systems that don’t support Docker or Singularity but can use Conda. Surprisingly, Conda works well beyond bioinformatics, especially given its strong support for Windows and macOS.

I had previously worked on an early version of [dockerthis](https://github.com/luciorq/dockerthis/tree/main), which was intended to serve as a backend for condathis, enabling Conda environments to run within containers. However, after seeing your presentation, I realized that `babelwhale` addresses many of the features I had envisioned. Your contributed function, `run_auto_mount()`, especially caught my attention, as I had spent considerable time conceptualizing a similar API.

Ultimately, my ideal goal is to create a unified API (resembling `processx::run()`) that allows seamless execution of CLI tools across different backends—be it WSL2, Docker, Apptainer, Singularity, Podman, or others. However, supporting such a diverse range of backends involves significant effort.

Thank you again for the inspiration! Your work has given me a lot to think about.

**Reply by joelnitta** [reply]
>  my ideal goal is to create a unified API (resembling processx::run()) that allows seamless execution of CLI tools across different backends

A worthy goal! Have you also considered nix? It ranks among the highest in terms of packages included. I wonder if it may be a good "backend". Also, would such a unified API also subsume `condathis`?

**Reply by luciorq** [reply]
## The Short Answer:

Yes, `nix` is definitely on our radar and has great potential as a "backend" for managing execution environments. If you’re already using `nix` to manage your system, you don’t need to change anything to start using `condathis`.

## The Long Answer:

`nix` excels in managing the layer "above" individual R installations and other system dependencies, but it may not be the most optimal choice for managing the steps of a workflow.

Without meaning to sound overly critical, I believe the way nix is often deployed can inadvertently hinder reproducibility over the long term. While it’s a powerful tool, it has certain limitations when the focus is on **both reproducibility and portability**. Here’s why:
- **Platform Constraints**: Like many Linux container backends (e.g., Docker, Singularity), nix requires modern Linux kernel versions to support features such as user namespaces and chroot-like functionality. This can pose challenges in shared environments, multi-user setups, and older HPC clusters—precisely the types of environments where scientific research often takes place.
- **Limited Windows Support**: Unlike conda, nix lacks robust support for Windows, further restricting its portability.

In contrast, I initially chose to focus on supporting conda packages due to their broader cross-platform compatibility:

- **Cross-Platform Ecosystem**: The most commonly used CLI tools and R, Python, and Perl libraries are well-supported on conda-forge. The data analysis, data science, and machine learning ecosystems, in particular, benefit from a rich selection of high-quality packages through conda-forge and adjacent channels.
- **Flexibility for Bioinformatics**: While bioinformatics tooling is generally more Linux/Unix-centric, the conda packaging format itself is inherently cross-platform. Creating packages with Windows compatibility is simply a matter of effort—anyone can develop a conda channel that prioritizes this.
- **Windows Advantages**: There are numerous command-line tools that work seamlessly on native Windows systems, giving conda an edge in certain cases, especially if you are writing R packages.

Looking ahead, WebAssembly’s emergence as a shared foundation for tools could lead to a convergence of these ecosystems, further blurring the lines between platform-specific solutions.

That said, I’d love to hear more about workflows where `nix` support shines. Could you point me to examples of tools that are better supported as NixPkgs than as conda packages? This is purely out of curiosity and a desire to explore areas where `nix` might offer unique advantages.

**Reply by luciorq** [reply]
> Also, would such a unified API also subsume condathis?

And yes, that is precisely the case.

**Comment by c1au6i0** [comment]
Great presentation, both of you! I just wanted to add one more thing about `condathis`. From a user point of view, a good thing is that it allows users to focus entirely on `R` without worrying about installing and setting up other tools, not even `conda`. Despite the name, the package automatically installs `micromamba`, which is self-contained and does not depend on `Python`.
Best,
